devlib/module/sched.py

```diff
@@ -0,0 +1,262 @@
+        self.domains = {}
+
+        for domain, fields in self.procfs.items():
+            self.domains[int(domain.lstrip("domain"))] = SchedDomain(fields)
+
+    def __init__(self, target, cpu):
+        self.procfs = target.read_tree_values(
+            target.path.join(self.sched_domain_root, 'cpu{}'.format(cpu)),
+            depth=5
+        )
+
+        self._build_domains()
+
+
+class SchedProcFSNode(object):
+    re_node = re.compile(r"(?P<name>.*)(?P<digits>\d+)$")
+
+    @staticmethod
+    def _ends_with_digits(node):
+        if not isinstance(node, basestring):
+            return False
+
+        return re.search(SchedProcFSNode.re_node, node) != None
+
+    @staticmethod
+    def _node_digits(node):
+        """
+        :returns: The ending digits of the procfs node
+        """
+        return int(re.search(SchedProcFSNode.re_node, node).group("digits"))
+
+    @staticmethod
+    def _node_name(node):
+        """
+        :returns: The name of the procfs node
+        """
+        return re.search(SchedProcFSNode.re_node, node).group("name")
+
+    @staticmethod
+    def _packable(node, entries):
+        """
+        :returns: Whether it makes sense to pack a node into a common entry
+        """
+        return (SchedProcFSNode._ends_with_digits(node) and
+                any([SchedProcFSNode._ends_with_digits(x) and
+                     SchedProcFSNode._node_digits(x) != SchedProcFSNode._node_digits(node) and
+                     SchedProcFSNode._node_name(x) == SchedProcFSNode._node_name(node)
+                     for x in entries]))
+
+    def _build_node(self, node):
+        if isinstance(node, dict):
+            return SchedProcFSNode(node)
+        else:
+            value = node
+
+            # Most nodes just contain numerical data, try to convert
+            try:
+                value = int(node)
+            except ValueError:
+                pass
+
+            return value
+
+    def __getattr__(self, name):
+        return self._dyn_attrs[name]
+
+    def __init__(self, nodes):
+        # First, reduce the procs fields by packing them if possible
+        # Find which entries can be packed into a common entry
+        packables = {
+            node : SchedProcFSNode._node_name(node) + "s"
+            for node in list(nodes.keys()) if SchedProcFSNode._packable(node, list(nodes.keys()))
+        }
+
+        self._dyn_attrs = {}
+
+        for dest in set(packables.values()):
+            self._dyn_attrs[dest] = {}
+
+        # Pack common entries
+        for key, dest in packables.items():
+            self._dyn_attrs[dest][SchedProcFSNode._node_digits(key)] = self._build_node(nodes[key])
+
+        # Build the other nodes
+        for key in nodes.keys():
+            if key in packables:
+                continue
+
+            self._dyn_attrs[key] = self._build_node(nodes[key])
+
+
+class SchedDomain(SchedProcFSNode):
+
+    # Domain flags obtained from include/linux/sched/topology.h on v4.17
```

Comment:
Should we add some sort of sanity check mechanism to verify that the target's kernel matches a min,max version?
Could eventually be done in the future... but maybe splitting out a warning in case the kernel is not a v4.17 just to say that we are using the flags as per their definition in kernel v4.17?